# Week 7 — Strings and Formatting

**COMP 1110 — Introduction to Computer Programming**

## + f-strings and Output Formatting

**Tuesday, October 20 & Thursday, October 22, 2026**

Instructor: **Quan Nguyen** · Room **OM 1241** · 9:30–11:20 AM

*Textbook: Python for Everybody, Chapter 6*

## Where we are

- **Last week (Week 6):** Midterm 1 — variables, conditionals, functions, loops (Ch. 1–5).
- **This week (Week 7):** Strings as *sequences* — indexing, slicing, methods — and how to make
  your output look professional with **f-strings** and format specifiers.
- **Next week (Week 8):** Files — reading a file line by line, and testing & documentation.

> Everything you read from a file next week arrives as a **string**. This week is the toolkit.

## Today's outcomes

By the end of these two classes you can:

1. Index, slice, and traverse a string with a loop, and explain why strings are **immutable**.
   `[CLO 1]`
2. Use `in`, string comparison, and string **methods** (`upper`, `strip`, `find`, `replace`,
   `split`) to search and clean text. `[CLO 3]`
3. **Parse** a piece of text to pull out the part you actually need. `[CLO 5]`
4. Format output with **f-strings**: field width, alignment, and rounding to decimal places.
   `[CLO 1]`
5. Recognise `IndexError` and `TypeError` from string code and fix them. `[CLO 4]`

## Reading

### [Chapter 6 — Strings](https://www.py4e.com/html3/06-strings)

Read it with VS Code open — **retype every example**. Change a value,
re-run it, see what breaks.

Focus on:

- 6.1–6.4 indexing, `len`, traversal, slices
- 6.5–6.7 immutability, counting, the `in` operator
- 6.10 string methods · 6.11 parsing
- 6.12 format operator (we will modernise this to f-strings)

**Worksheet 5 was due last night at 11:59 PM.** Worksheet 6 goes live at 9:00 this morning.

## A string is a *sequence* of characters

A string is not one indivisible blob. It is an ordered row of characters, each with a numbered
position — its **index**.

```
 s = "Python"

   P    y    t    h    o    n
   0    1    2    3    4    5
  -6   -5   -4   -3   -2   -1
```

Use square brackets to pull one character out: `s[0]`.

**Counting starts at 0.** The first character is at index `0`, not `1`.

In [1]:
course = "Python"

print(course[0])
print(course[1])
print(course[5])

P
y
n


### Negative indexes count from the right

`s[-1]` is the **last** character. This is handy because you do not need to know the length.

In [2]:
course = "Python"

print(course[-1])   # last character
print(course[-2])   # second-last
print(course[-6])   # same as course[0]

n
o
P


## `len()` — how long is it?

`len(s)` gives the number of characters in the string.

The **last valid index is `len(s) - 1`**, because indexes start at 0. Forgetting the `- 1` is the
single most common string bug in this course.

In [3]:
message = "Good morning"

print(len(message))          # 12 characters (the space counts!)
print(message[len(message) - 1])   # last character

12
g


In [4]:
message = "Good morning"
print(message[len(message)])   # one past the end

IndexError: string index out of range

`IndexError: string index out of range` — index 12 does not exist in a 12-character string.
Valid indexes are `0` through `11`.

## Predict the output #1

No running Python. Write your answer down first.

```python
word = "temperature"
print(word[0])
print(word[-1])
print(word[3:7])
print(word[:4])
print(len(word[2:5]))
```

In [5]:
word = "temperature"
print(word[0])
print(word[-1])
print(word[3:7])
print(word[:4])
print(len(word[2:5]))

t
e
pera
temp
3


**Why:** `t` · `e` · `pera` (indexes 3,4,5,6 — 7 is excluded) · `temp` (start defaults to 0) ·
`3` (`word[2:5]` is `"mpe"`, three characters — `5 - 2`).

**Rule:** a slice `s[a:b]` has `b - a` characters when both are in range.

## Traversal: visiting every character

A very common pattern: do something once *per character*.

Two ways —

1. **With an index and a `while` loop** — you know *where* you are.
2. **With a `for` loop** — shorter, and you cannot fall off the end.

In [6]:
fruit = "banana"

index = 0
while index < len(fruit):
    print(index, fruit[index])
    index = index + 1

0 b
1 a
2 n
3 a
4 n
5 a


### The same traversal with `for`

No counter, no `len`, no off-by-one risk. Prefer this when you do not need the index.

In [7]:
fruit = "banana"

for letter in fruit:
    print(letter)

b
a
n
a
n
a


## Slices: taking a piece

`s[start:end]` gives the characters from `start` **up to but not including** `end`.

| Slice | Meaning |
| :--- | :--- |
| `s[2:5]` | indexes 2, 3, 4 |
| `s[:5]` | from the beginning to index 4 |
| `s[5:]` | from index 5 to the end |
| `s[:]` | the whole string |
| `s[-3:]` | the last three characters |

If `start >= end`, you get the **empty string** `''` — not an error.

In [8]:
date_text = "2026-10-20"

print(date_text[0:4])    # year
print(date_text[5:7])    # month
print(date_text[8:])     # day
print(date_text[-2:])    # day again, counting from the right

2026
10
20
20


In [9]:
date_text = "2026-10-20"

print(repr(date_text[7:3]))    # start after end -> empty
print(repr(date_text[4:200]))  # end past the end is fine, no error

''
'-10-20'


Slices are **forgiving**; single indexes are **not**. `date_text[200]` raises `IndexError`,
but `date_text[4:200]` just stops at the end of the string.

## Strings are **immutable**

You cannot change a character of an existing string. Once a string object exists, its characters
are fixed forever.

```python
greeting = "Hello, world"
greeting[0] = "J"      # ← TypeError
```

This is not a limitation to work around — it is a guarantee. If you pass a string to a function,
that function cannot secretly alter it.

In [10]:
greeting = "Hello, world"
greeting[0] = "J"

TypeError: 'str' object does not support item assignment

`TypeError: 'str' object does not support item assignment`

**The fix: build a new string** instead of editing the old one.

In [11]:
greeting = "Hello, world"
new_greeting = "J" + greeting[1:]

print(new_greeting)
print(greeting)          # the original is untouched

Jello, world
Hello, world


Every string "change" in Python is really *making a new string*. `upper()`, `strip()`, `replace()`
all **return** a new string and leave the original alone.

## Looping and counting

The counter pattern from Week 5, applied to text:

1. Set a counter to `0` **before** the loop.
2. Look at each character.
3. Add `1` when it matches.
4. Report **after** the loop.

This is the shape of almost every "how many …?" question about text.

In [12]:
tweet = "the rain in spain stays mainly in the plain"

count = 0
for letter in tweet:
    if letter == "a":
        count = count + 1

print("Number of a's:", count)

Number of a's: 5


### Wrap it in a function — reusable and testable

Week 4 said: if you did it twice, make it a function.

In [13]:
def count_letter(text, target):
    """Return how many times target appears in text."""
    count = 0
    for letter in text:
        if letter == target:
            count = count + 1
    return count


print(count_letter("mississippi", "s"))
print(count_letter("mississippi", "z"))

4
0


## Predict the output #2

```python
def mystery(text):
    total = 0
    for letter in text:
        if letter in "aeiou":
            total = total + 1
    return total


print(mystery("programming"))
print(mystery("RHYTHM"))
```

In [14]:
def mystery(text):
    total = 0
    for letter in text:
        if letter in "aeiou":
            total = total + 1
    return total


print(mystery("programming"))
print(mystery("RHYTHM"))

3
0


**Why:** `3` — `o`, `a`, `i` in "programming". Then `0` — the vowels in `"RHYTHM"` are absent, and
even if there were an `A` it is uppercase, so `"A" in "aeiou"` would be `False`.

**Lesson:** comparisons are case-sensitive. Real code calls `.lower()` first.

## The `in` operator

`in` is a boolean operator that asks: *is this string a **substring** of that one?*

```python
"a" in "banana"        # True
"seed" in "banana"     # False
"nan" in "banana"      # True  -- substrings, not just single characters
```

It returns `True` or `False`, so it goes straight into an `if`.

In [15]:
subject = "URGENT: your invoice is overdue"

if "urgent" in subject.lower():
    print("Flagging this message.")
else:
    print("Normal message.")

Flagging this message.


## Comparing strings

`==` tests equality. `<` and `>` compare **alphabetically** — more precisely, by character code.

That means **all uppercase letters sort before all lowercase letters**: `"Zebra" < "apple"` is
`True`, because `Z` (code 90) comes before `a` (code 97).

To compare fairly, normalise case first with `.lower()`.

In [16]:
name = "Charlie"

if name < "M":
    print(name, "is in the first half of the alphabet")
else:
    print(name, "is in the second half")

print("Zebra" < "apple")
print("Zebra".lower() < "apple".lower())

Charlie is in the first half of the alphabet
True
False


## String **methods**

A method is a function that belongs to a value. You call it with a dot:

```python
text.upper()
```

Every string method **returns a new string** — remember, strings are immutable.

Use `dir(str)` to list them all, and `help(str.strip)` to read one.

In [17]:
raw = "   Ada Lovelace\n"

print(repr(raw.upper()))
print(repr(raw.lower()))
print(repr(raw.strip()))         # removes whitespace from BOTH ends
print(repr(raw))                 # unchanged!

'   ADA LOVELACE\n'
'   ada lovelace\n'
'Ada Lovelace'
'   Ada Lovelace\n'


### Searching: `startswith`, `endswith`, `find`
`find` returns the **index** of the first match, or **`-1`** if there is none — not an error.

In [18]:
line = "From stephen.marquard@uct.ac.za Sat Jan  5 09:14:16 2008"

print(line.startswith("From"))
print(line.endswith("2008"))
print(line.find("@"))          # index of the first @
print(line.find("zebra"))      # not found -> -1
print(line.find("a", 10))      # start searching at index 10

True
True
21
-1
14


### Changing: `replace`
`replace(old, new)` returns a new string with **every** occurrence swapped.

In [19]:
notice = "Class is cancelled on Monday. Monday only."

print(notice.replace("Monday", "Tuesday"))
print(notice.replace("Monday", "Tuesday", 1))   # only the first one
print(notice)                                    # original untouched

Class is cancelled on Tuesday. Tuesday only.
Class is cancelled on Tuesday. Monday only.
Class is cancelled on Monday. Monday only.


### Splitting: `split`

`split()` cuts a string into pieces at the whitespace and hands back a **list** of them.

We meet lists properly in **Week 9**. For today, all you need: the pieces come back numbered from
`0`, and you can grab one with `[ ]`, just like a string index.

In [20]:
line = "Bianca Reyes 87 92 78"

parts = line.split()
print(parts)
print(parts[0])       # first name
print(parts[1])       # last name
print(len(parts))     # how many pieces

['Bianca', 'Reyes', '87', '92', '78']
Bianca
Reyes
5


## Parsing a string

**Parsing** = finding the piece of text you actually want inside a larger piece.

The recipe:

1. Find a landmark with `find()`.
2. Slice relative to that landmark.
3. `strip()` off any stray whitespace.

Below: pull the **host name** out of an email log line — exactly the Chapter 6 example.

In [21]:
line = "From stephen.marquard@uct.ac.za Sat Jan  5 09:14:16 2008"

at_pos = line.find("@")
space_pos = line.find(" ", at_pos)
host = line[at_pos + 1:space_pos]

print(at_pos, space_pos)
print(host)

21 31
uct.ac.za


## Live coding #1 — together

**Problem.** Write `initials(full_name)` that turns `"ada lovelace"` into `"A.L."`.

Rules: the name has exactly two words separated by a space; the result is the two capitalised
first letters, each followed by a period.

Think out loud before typing: where is the space? what do I slice?

In [22]:
# Live coding #1
# initials("ada lovelace")  ->  "A.L."
# initials("Grace Hopper")  ->  "G.H."

In [23]:
def initials(full_name):
    """Return the two initials of a two-word name, e.g. 'A.L.'."""
    space_pos = full_name.find(" ")
    first = full_name[0].upper()
    last = full_name[space_pos + 1].upper()
    return first + "." + last + "."


print(initials("ada lovelace"))
print(initials("Grace Hopper"))

A.L.
G.H.


# Part 2 — Output formatting

## Making output look professional

So far our output has looked like this:

```
Total: 12.666666666666666
```

Nobody ships that. Today we fix it — permanently.

The tool: the **f-string**.

## f-strings

Put an `f` before the opening quote. Anything inside `{ }` is a **Python expression**, evaluated
and dropped into the text.

```python
name = "Sam"
print(f"Hello, {name}!")
```

Compare with what we have been doing:

```python
print("Hello, " + name + "!")        # concatenation — breaks on numbers
print("Hello,", name)                 # commas — you cannot control spacing
```

**From this week on, use f-strings.**

In [24]:
name = "Sam"          # imagine the user typed this
score = 87
total = 100

print(f"Hello, {name}!")
print(f"{name} scored {score} out of {total}.")
print(f"That is {score / total * 100} percent.")

Hello, Sam!
Sam scored 87 out of 100.
That is 87.0 percent.


### Expressions inside the braces

Anything that produces a value works: arithmetic, method calls, function calls.

In [25]:
name = "sam"
score = 87

print(f"{name.upper()} got {score} marks.")
print(f"Next mark up: {score + 1}")
print(f"Name length: {len(name)}")
print(f"Passed: {score >= 50}")

SAM got 87 marks.
Next mark up: 88
Name length: 3
Passed: True


## Rounding to decimal places

Inside the braces, add `:` then a **format specification**.

`{value:.2f}` means *fixed-point, 2 digits after the decimal*.

This is how you print **money**. Always 2 decimals — `4.5` should display as `4.50`.

In [26]:
subtotal = 19.99
tax_rate = 0.12
tax = subtotal * tax_rate
total = subtotal + tax

print(f"Tax:   {tax}")            # ugly
print(f"Tax:   ${tax:.2f}")       # money
print(f"Total: ${total:.2f}")
print(f"Rate:  {tax_rate:.1%}")   # percent, 1 decimal

Tax:   2.3987999999999996
Tax:   $2.40
Total: $22.39
Rate:  12.0%


`.2f` **rounds for display only** — the variable itself keeps its full precision. Never round your
stored value; round when you print.

## Field width and alignment

`{value:10}` reserves **10 characters** of space for the value.

| Spec | Meaning |
| :--- | :--- |
| `{x:10}` | width 10 — text left-aligns, numbers right-align |
| `{x:<10}` | force **left** align |
| `{x:>10}` | force **right** align |
| `{x:^10}` | **centre** |
| `{x:>8.2f}` | width 8, 2 decimals, right-aligned |
| `{x:,}` | thousands separators: `1,250,000` |

Width + alignment is how you build a **table** that lines up.

In [27]:
label = "Mouse"

print(f"[{label:<12}]")
print(f"[{label:>12}]")
print(f"[{label:^12}]")
print(f"[{1250000:,}]")
print(f"[{3.14159:>9.3f}]")

[Mouse       ]
[       Mouse]
[   Mouse    ]
[1,250,000]
[    3.142]


### Putting it together: an aligned receipt

Names left, money right, decimals fixed. Notice how the header uses the *same* widths.

In [28]:
print(f"{'Item':<14}{'Qty':>4}{'Amount':>10}")
print("-" * 28)
print(f"{'Wireless Mouse':<14}{2:>4}{2 * 24.99:>10.2f}")
print(f"{'USB-C Cable':<14}{3:>4}{3 * 9.50:>10.2f}")
print(f"{'Laptop Stand':<14}{1:>4}{1 * 129.00:>10.2f}")
print("-" * 28)
print(f"{'TOTAL':<14}{6:>4}{207.48:>10.2f}")

Item           Qty    Amount
----------------------------
Wireless Mouse   2     49.98
USB-C Cable      3     28.50
Laptop Stand     1    129.00
----------------------------
TOTAL            6    207.48


## Predict the output #3

Count the characters carefully — dots mark where spaces land.

```python
value = 7.456
print(f"[{value:.1f}]")
print(f"[{value:8.2f}]")
print(f"[{'hi':<5}|]")
print(f"[{42:>6}]")
```

In [29]:
value = 7.456
print(f"[{value:.1f}]")
print(f"[{value:8.2f}]")
print(f"[{'hi':<5}|]")
print(f"[{42:>6}]")

[7.5]
[    7.46]
[hi   |]
[    42]


**Why:**

- `[7.5]` — one decimal, rounded up.
- `[    7.46]` — total width **8**, so `7.46` (4 chars) gets 4 leading spaces.
- `[hi   |]` — width 5, left-aligned, 3 trailing spaces before the `|`.
- `[    42]` — width 6, right-aligned by default for numbers anyway.

## `.format()` — the previous generation

You will see this in older code and in the textbook. Same specifiers **after the colon**, but the
values go in the argument list instead of inline.

```python
"Total: {:.2f}".format(total)
```

It still works, and it is useful when the template string is built separately from the values.
For everything you write, **prefer f-strings** — the value is right where you read it.

In [30]:
name = "Sam"
total = 208.4832

print("{} owes ${:.2f}".format(name, total))
print("{0} owes ${1:.2f} — thanks, {0}!".format(name, total))
print(f"{name} owes ${total:.2f}")          # same thing, clearer

Sam owes $208.48
Sam owes $208.48 — thanks, Sam!
Sam owes $208.48


## The `%` operator — legacy, recognise but do not write

The oldest of the three. Chapter 6 calls it the *format operator*.

```python
"%d items cost %.2f dollars" % (3, 74.97)
```

`%d` = integer · `%s` = string · `%.2f` = float with 2 decimals. The values come after the `%`
in parentheses.

**Know it so you can read old code.** Write f-strings.

In [31]:
count = 3
amount = 74.97

print("%d items cost %.2f dollars" % (count, amount))
print("Hello, %s!" % "Sam")
print(f"{count} items cost {amount:.2f} dollars")   # what we write today

3 items cost 74.97 dollars
Hello, Sam!
3 items cost 74.97 dollars


## Live coding #2 — together

**Problem.** A line of text holds a student's record:

```
"Bianca Reyes|88|92|79"
```

Write `report(line)` that prints one aligned row:

```
Bianca Reyes        avg  86.33
```

Name left-aligned in 20 columns; the average to 2 decimals. Use `split("|")` and `float()`.

In [32]:
# Live coding #2
# report("Bianca Reyes|88|92|79")
#   ->  Bianca Reyes        avg  86.33
record = "Bianca Reyes|88|92|79"

In [33]:
def report(line):
    """Print an aligned name-and-average row from a pipe-separated record."""
    parts = line.split("|")
    name = parts[0]
    average = (float(parts[1]) + float(parts[2]) + float(parts[3])) / 3
    print(f"{name:<20}avg {average:>6.2f}")


report("Bianca Reyes|88|92|79")
report("Wei Zhang|100|95|91")

Bianca Reyes        avg  86.33
Wei Zhang           avg  95.33


## Common errors this week (1 of 2)

**1. `IndexError: string index out of range`**
You used index `len(s)`. The last index is `len(s) - 1`.

**2. `TypeError: 'str' object does not support item assignment`**
You wrote `word[0] = "J"`. Strings are immutable — build a new string:
`word = "J" + word[1:]`.

**3. The method ran but nothing changed.**
`text.upper()` on its own throws the result away. Methods **return** a new string:
`text = text.upper()`.

In [34]:
name = "sam"
name.upper()          # result discarded!
print(name)

name = name.upper()   # this is what you meant
print(name)

sam
SAM


## Common errors this week (2 of 2)

**4. `find()` returned `-1` and you sliced with it.**
No error is raised — you silently get the wrong text. Always check:

```python
pos = line.find("@")
if pos == -1:
    print("No @ in this line")
else:
    ...
```

**5. Forgetting the `f`.** `print("Total: {total:.2f}")` prints the braces literally.
No error, just wrong output. If you see `{` in your output, you dropped the `f`.

In [35]:
total = 12.5

print("Total: {total:.2f}")     # forgot the f — prints the braces
print(f"Total: {total:.2f}")    # correct

Total: {total:.2f}
Total: 12.50


## Summary

- A string is a **sequence**: `s[0]` is the first character, `s[-1]` the last,
  `len(s) - 1` the last valid index.
- **Slices** `s[a:b]` include `a`, exclude `b`, and never raise `IndexError`.
- Strings are **immutable** — every "change" returns a **new** string. Assign it or lose it.
- `in` tests substrings; `<` compares alphabetically and is **case-sensitive** — `.lower()` first.
- Methods to know: `upper`, `lower`, `strip`, `startswith`, `find` (returns `-1`!), `replace`,
  `split`.
- **f-strings** are the way we format from now on: `{value:.2f}` for money,
  `{text:<14}` and `{n:>8}` for aligned tables.

## Worksheet 6

**Released:** Tuesday, October 20 at 9:00 AM · **Due:** Monday, October 26 at **11:59 PM**

It covers:

- Indexing and slicing — including negative indexes and boundary cases.
- A counting-traversal function (count characters matching a condition).
- Parsing: pull a field out of a log line using `find()` and slicing.
- Formatting: print a small aligned table with prices rounded to 2 decimals.
- One **trace-and-explain**: given string code, state its exact output.

60–90 minutes if you have done the reading. Start it as soon as it is released.

## Next class — Week 8

**Tue Oct 27:** **Files** — `open()`, reading a file line by line, searching a file,
writing files, and surviving a missing file with `try` / `except`.

**Thu Oct 29:** **Testing & documentation** — test plans, edge cases, `assert`, a first look at
`unittest`, docstrings, and README files.

> Every line you read from a file is a **string** with a `\n` on the end. `strip()`, `split()`,
> and `find()` are about to become your daily tools. Bring today's notes.